### PR TITLE
mbedtls: Add CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN

### PR DESCRIPTION
**Pull request by pfalcon**
This value allows to set max payload length of a TLS protocol
message, and passed thru to mbedTLS as MBEDTLS_SSL_MAX_CONTENT_LEN
setting. The only safe value is 16384, which translates to 32KB
of RAM required just for mbedTLS input/output buffers. Any other
value can be configured *only* per a particular application
(e.g. knowing that it won't pass more than spefific amount of
data at once and/or won't connect to a server with a long cert
chain). Previosuly, we had quite an adhoc and inflexible config
with random values for that setting, based on protocol.

Note that while the safe value is 16384, "backward compatible"
default of 1500 is used (good for DTLS on the other hand).

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
Further clean up along the lines of https://github.com/zephyrproject-rtos/zephyr/pull/6219, https://github.com/zephyrproject-rtos/zephyr/pull/6221

**Comment by mike-scott** [comment]
@pfalcon This looks very similar to the CoAP specific commit: https://github.com/zephyrproject-rtos/zephyr/commit/2fc554b8fcd37e589448010ff4dfab4affff2af4

I wonder if we should remove the extra config from the CoAP lib and re-use this config?

**Comment by mike-scott** [comment]
I wonder if we could remove the coap specific config in favor of this "mini" config which could be #ifdef'd for features requiring TCP(TLS) / UDP(DTLS).  So that if TCP isn't selected it would be the equivalent of the coap config for DTLS.

**Comment by pfalcon** [comment]
> @pfalcon This looks very similar to the CoAP specific commit: 2fc554b
> I wonder if we should remove the extra config from the CoAP lib and re-use this config?

Thanks for the pointer, that definitely would be in the scope of this PR, I'll look into adding it.

**Comment by pfalcon** [comment]
>  So that if TCP isn't selected it would be the equivalent of the coap config for DTLS.

I had an idea like, because I'd really like to have a default of 16384 ;-). Then, if TCP is not selected, we can have a default of 1500.

That's all a bit brave of course without changing mbedTLS itself (e.g. https://github.com/ARMmbed/mbedtls/pull/1368). I've marked this as 1.12, so we have some time to think it all.

**Comment by mike-scott** [comment]
@pfalcon correction on my part: I believe that config-coap.h is a trimmed down version of config-mini-dtls1_2.h (though the diff looks messy).

**Comment by jukkar** [comment]
Adding this new config option llooks very reasonable so +1 from me for this one.

**Comment by pfalcon** [comment]
@mike-scott, @jukkar : I've refactored COAP-specific setting to use the generic one instead. In my list, this should be ready to merge. Please re-review.

File: ext/lib/crypto/mbedtls/configs/config-coap.h [hunk 1]
```diff
@@ -91,15 +91,11 @@
...
+ * Allow to save RAM at the expense of interoperability: do this only if you
+ * control both ends of the connection!  (See comments in "mbedtls/ssl.h".)
  * The optimal size here depends on the typical size of records.
  */
-#ifndef CONFIG_COAP_MBEDTLS_SSL_MAX_CONTENT_LEN
-#define MBEDTLS_SSL_MAX_CONTENT_LEN	1500
-#else
-#define MBEDTLS_SSL_MAX_CONTENT_LEN	CONFIG_COAP_MBEDTLS_SSL_MAX_CONTENT_LEN
-#endif
+#define MBEDTLS_SSL_MAX_CONTENT_LEN  CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN
```
Comment (mike-scott) [inline ext/lib/crypto/mbedtls/configs/config-coap.h]: I know we have CONFIG_MBEDTLS_BUILTIN=y all over our sample .conf files at the moment, but CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN depends on CONFIG_MBEDTLS_BUILTIN and it's still a manual setting.  Do we want to remove the #ifndef here, or merely change it to CONFIG_MBEDTLS_SSL_MAX_CONTENT_LEN?

Comment (pfalcon) [reply ext/lib/crypto/mbedtls/configs/config-coap.h]: Well, my idea is that people either: a) use CONFIG_MBEDTLS_BUILTIN=y, choose one of (hopefully growing number of) Zephyr-provided mbedTLS configs, and are beyond that, able to adjust "interesting" options using Zephyr Kconfig options, or b) use CONFIG_MBEDTLS_BUILTIN=n and used both mbedTLS code and config coming from outside Zephyr. IMHO, that's reasonable.

If you have concerns specifically about config-coap.h, I can put ifdef back.

Comment (mike-scott) [reply ext/lib/crypto/mbedtls/configs/config-coap.h]: Is there a case for people to use a Zephyr provided config (like coap.h) but building from an external source.  If we don't think we should be supporting that then I'm ok with it as is.  If not, we probably need the #ifndef back.

Comment (pfalcon) [reply ext/lib/crypto/mbedtls/configs/config-coap.h]: My opinion is like above - it makes sense to support "polar" options, either fully integrated into Zephyr, or fully external. I guess people would use external option to retain full control of how mbedTLS is built. Using Z's config won't allow that, as Z's config may evolve with it ("unexpectedly" to users). Then trying to support various mix-in and half-way options only leads to extra effort (all those ifdef's) and confusion.

Anyway, that's just my opinions. @jukkar, @tbursztyka, @nashif : Any thoughts on the discussion above?

Comment (mike-scott) [reply ext/lib/crypto/mbedtls/configs/config-coap.h]: @pfalcon thanks for the discussion.  I'm ok w/ leaving it as is.

**Review by mike-scott — APPROVED** [review]
LGTM

**Review by nashif — APPROVED** [review]

**Pull request merged**